### PR TITLE
Add DNS capture into CNI.

### LINKED ISSUE #29511: DNS redirect iptables rules not created by istio-cni
(**NOTE**: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>)

**Bug description**

We are trying to create alternate hostname with ServiceEntry for a service and use it to access the service. We are using istio-cni plugin to install the iptable rules for traffic redirection. We have enabled istio dns in sidecars via ISTIO_META_DNS_CAPTURE and ISTIO_META_PROXY_XDS_VIA_AGENT flag. However it seems that istio-cni is not installing udp rules to redirect the dns lookup to the sidecar.

I did a quick grep in the cni directory for port 15053 but could not find any match on any source file. Is the changes implemented for cni to create udp rules for dns?

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Expected behavior**
DNS traffic needs to be captured by sidecar and resolve the same.

**Steps to reproduce the bug**
1. Install istio with cni. We used the following istio operator file :-
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: example-istiocontrolplane
spec:
  profile: demo
  meshConfig:
    defaultConfig:
      proxyMetadata:
        ISTIO_META_DNS_CAPTURE: "true"
        ISTIO_META_PROXY_XDS_VIA_AGENT: "true"
  components:
    cni:
      enabled: true
  values:
    cni:
      excludeNamespaces:
       - istio-system
       - kube-system
      logLevel: info
```
2. Create a service (We used `samples/httpbin/httpbin.yaml`)
3. Create a service entry. We used following :-
```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: httpbin-test1-global
spec:
  addresses:
  - 10.109.229.252
  endpoints:
  - address: httpbin.test.svc.cluster.local
    ports:
      http: 8000
  hosts:
  - httpbin.test.global
  exportTo:
  - "*"
  location: MESH_INTERNAL
  ports:
  - name: http
    number: 8000
    protocol: http
  resolution: DNS
```
3. Try to access the service from a client pod :-
 ```
$ curl httpbin.test.global:8000
curl: (6) Could not resolve host: httpbin.test.global
/ $ curl httpbin.test.global:8000/headers
curl: (6) Could not resolve host: httpbin.test.global
/ $ command terminated with exit code 6
```
4.  Check the iptable. rules :-
```
docker exec 68194b20bcae ip netns exec cni-c623dbf9-7ac4-e2ab-e032-24d991970733 iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination
ISTIO_INBOUND  tcp  --  anywhere             anywhere

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
ISTIO_OUTPUT  tcp  --  anywhere             anywhere

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination

Chain ISTIO_INBOUND (1 references)
target     prot opt source               destination
RETURN     tcp  --  anywhere             anywhere             tcp dpt:15008
RETURN     tcp  --  anywhere             anywhere             tcp dpt:22
RETURN     tcp  --  anywhere             anywhere             tcp dpt:15020
RETURN     tcp  --  anywhere             anywhere             tcp dpt:15021
RETURN     tcp  --  anywhere             anywhere             tcp dpt:15090
ISTIO_IN_REDIRECT  tcp  --  anywhere             anywhere

Chain ISTIO_IN_REDIRECT (3 references)
target     prot opt source               destination
REDIRECT   tcp  --  anywhere             anywhere             redir ports 15006

Chain ISTIO_OUTPUT (1 references)
target     prot opt source               destination
RETURN     tcp  --  anywhere             anywhere             tcp dpt:15020
RETURN     all  --  127.0.0.6            anywhere
ISTIO_IN_REDIRECT  all  --  anywhere            !localhost            owner UID match 1337
RETURN     all  --  anywhere             anywhere             ! owner UID match 1337
RETURN     all  --  anywhere             anywhere             owner UID match 1337
ISTIO_IN_REDIRECT  all  --  anywhere            !localhost            owner GID match 1337
RETURN     all  --  anywhere             anywhere             ! owner GID match 1337
RETURN     all  --  anywhere             anywhere             owner GID match 1337
RETURN     all  --  anywhere             localhost
ISTIO_REDIRECT  all  --  anywhere             anywhere

Chain ISTIO_REDIRECT (1 references)
target     prot opt source               destination
REDIRECT   tcp  --  anywhere             anywhere             redir ports 15001
```

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)
 istioctl version --remote
client version: 1.8.0
control plane version: 1.8.0
data plane version: 1.8.0 (4 proxies)

**How was Istio installed?**
Istio was installed via operator :-
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: example-istiocontrolplane
spec:
  profile: demo
  meshConfig:
    defaultConfig:
      proxyMetadata:
        ISTIO_META_DNS_CAPTURE: "true"
        ISTIO_META_PROXY_XDS_VIA_AGENT: "true"
  components:
    cni:
      enabled: true
  values:
    cni:
      excludeNamespaces:
       - istio-system
       - kube-system
      logLevel: info
```
**Environment where the bug was observed (cloud vendor, OS, etc)**
We reproduced on a kind cluster.

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

### PR DESCRIPTION
**Pull request by bianpengyuan**
fix #29511

### CODE REVIEW

File: cni/cmd/istio-cni/redirect.go [hunk 1]
```diff
@@ -246,12 +248,19 @@ func NewRedirect(annotations map[string]string) (*Redirect, error) {
...
 	redir.excludeInboundPorts += "15020,15021,15090"
 	redir.excludeInboundPorts = strings.Join(dedupPorts(splitPorts(redir.excludeInboundPorts)), ",")
-	isFound, redir.kubevirtInterfaces, valErr = getAnnotationOrDefault("kubevirtInterfaces", annotations)
+	isFound, redir.kubevirtInterfaces, valErr = getAnnotationOrDefault("kubevirtInterfaces", pi.Annotations)
 	if valErr != nil {
 		log.Errorf("Annotation value error for value %s; annotationFound = %t: %v",
 			"kubevirtInterfaces", isFound, valErr)
 		return nil, valErr
 	}
+	if v, found := pi.ProxyConfig.ProxyMetadata[options.DnsCaptureByAgent.Name]; found {
```
Comment (howardjohn) [inline cni/cmd/istio-cni/redirect.go:257]: ProxyConfig may be more generally useful, but it seems we only use it for ProxyMetadata. Technically the agent doesn't read ProxyMetadata, it reads the env, which can be set by other means. Should we read the env possibly?

If we want the proxy config struct for other purposes we could "backfill" the env from the pod spec into ProxyConfig.ProxyMetadata I guess.

This should work in 95% of cases as is, just thinking about less common cases

**Review by howardjohn — COMMENTED** [review]
Looks great

File: cni/cmd/istio-cni/redirect.go (hunk 1 above)
Comment (bianpengyuan) [reply cni/cmd/istio-cni/redirect.go:257]: > ProxyConfig may be more generally useful, but it seems we only use it for ProxyMetadata.

Yeah I am thinking about reading include/exclude ip/port from proxyConfig, and avoid adding those annotations in sidecar injection templates.

>  Technically the agent doesn't read ProxyMetadata, it reads the env, which can be set by other means. Should we read the env possibly?

Yeah I am a bit hesitant to read from individual env variable, and I guess to make DNS capture beta, we will need some better first class API? But anyway I guess reading from env var is the most correct way to do it right now.

So for this change, I will keep ProxyConfig change for other usages, but read DNS capture information from env, wdyt?

Comment (howardjohn) [reply cni/cmd/istio-cni/redirect.go:257]: >  I am thinking about reading include/exclude ip/port from proxyConfig

Do we have those currently there? or in the future?

SGTM

Comment (bianpengyuan) [reply cni/cmd/istio-cni/redirect.go:257]: > Do we have those currently there? or in the future?

Ohh, they are just values.. Somehow I thought they were proxyConfig. Yeah, I guess one day in the future then :). Then I guess one thing we can clean up from sidecar template is interceptionMode, which is available at proxyConfig. Anyway, that's work for later.

File: cni/pkg/install-cni/cmd/root.go [hunk 2]
```diff
@@ -151,5 +153,12 @@ func constructConfig() (*config.Config, error) {
 		}
 	}
 
+	// Get install-cni DNS server addresses
+	dnsConfig, err := dns.ClientConfigFromFile("/etc/resolv.conf")
```
Comment (howardjohn) [inline cni/pkg/install-cni/cmd/root.go:157]: I am not sure this is right. It should be from the pod, not CNI. They may not be the same... but I am not sure if we can do any better here either. 

Maybe parse `pod.spec.dnsConfig.nameservers`? seems flimsy though, I think the image can set them arbitrarily as well

Comment (bianpengyuan) [reply cni/pkg/install-cni/cmd/root.go:157]: Are you suggesting parsing `dnsConfig.nameservers` first, and fall back to install-cni DNS server if not specified? Looking at the istio-cni log, I don't think dnsConfig is included in the pod spec as long as user does not specify it.

Comment (howardjohn) [reply cni/pkg/install-cni/cmd/root.go:157]: I think `dnsConfig.nameservers` will override the defaults, so I think we would fallback to install-cni. But not 100% sure, or if there are other ways to modify /etc/resolv.conf (like modifying it in the image)

Comment (bianpengyuan) [reply cni/pkg/install-cni/cmd/root.go:157]: Or make CNI DNS capture all request to 53. I know there is an issue with infinite loop due to local dns cache, but not sure if that is often an issue for k8s?

Comment (howardjohn) [reply cni/pkg/install-cni/cmd/root.go:157]: Its probably extremely unlikely on k8s, it only happens when you have another local dns proxy

**Comment by bianpengyuan** [comment]
@howardjohn Do you think if it is ok to make istio cni enablement a test framework flag? Local DNS cache test fails with CNI which we know won't work: https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32416/integ-pilot-k8s-tests_istio/1387096174833438720. We should just skip those cases when CNI is enabled, but that information is tricky to get without istio.enable_cni being a test framework flag.

**Review by howardjohn — COMMENTED** [review]
Its a more general problem really. We are starting to treat our tests more and more like "conformance tests" where we bring some environment and then run a bunch of tests against it. So we have things like RequireIstioVersion, RequireClusterVersion, --istio.test.revision, --istio.test.skipVM, etc. We are also making tests workload agnostic which is pretty similar

We probably want a more general solution long term where a test can declare its dependencies and a test-runner can declare how its setup.

Short term a flag is probably fine, or in the job that runs CNI just add `--istio.test.skip`?

@stevenctl for more opinions

**Comment by stevenctl** [comment]
I think a flag is fine; maybe we need some general flag like `--istio.test.features=-vm,+cni` where we skip tests that require VMs or skip tests that will break with CNI enabled. 

If we do add `--istio.test.cniEnabled`, I'm not sure we should have it control setting the install options, or if the flag should just help inform test skips.

**Comment by bianpengyuan** [comment]
Ok, thanks for the feedback. I will add a flag for now then.

> If we do add --istio.test.cniEnabled, I'm not sure we should have it control setting the install options, or if the flag should just help inform test skips.

I think we should just control all CNI related stuff with this flag if we have it. Long term, when we have a feature skipping flag, we can go back and use operator install options. I created an issue to track this: https://github.com/istio/istio/issues/32493

**Comment by bianpengyuan** [comment]
@howardjohn @stevenctl ptal thanks!

**Review by howardjohn — APPROVED** [review]

**Comment by neerajaustin** [comment]
Hello, I am waiting on this fix to be used in our product deployment. Please indicate which release version would this be available? I see istio 1.10 is released. So just wanted to get an idea.

**Comment by jbilliau-rcd** [comment]
Same here; we are waiting to implement the new SmartDNS feature until this is released; currently on 1.8.6

**Comment by bianpengyuan** [comment]
Unfortunately this did not meet 1.10 cut. Will be in 1.11.

**Comment by diranged** [comment]
> Unfortunately this did not meet 1.10 cut. Will be in 1.11.

Quick question - any chance we can either run a pre-release in our dev clusters to try it out, or you can shed some light on when 1.11 is scheduled for release?

**Comment by howardjohn** [comment]
https://github.com/istio/istio/wiki/Dev%20Builds describes accessing dev
builds

**Comment by diranged** [comment]
@howardjohn thanks i can confirm that on the 1.11-dev release, the istio-cni is properly setting up the DNS iptables rules.. thanks!!

**Comment by noah8713** [comment]
@howardjohn  @bianpengyuan  Can we also backport this to 1.10 too?

**Pull request merged**